WIP: refactoring and more checks in prince source

edoput opened the pull request:
This will try to not break prince and make it more robust

- under the directory `prince` (kind of ) everything is checked for bad malloc and kind of handled
- reading the configuration provides hints for mistypes and such
- logs to stderr and stdout are divided
- add cmake build definitions
- move headers to different directories
- install binaries and libraries to /usr/local by default
- source files with a high count of loc have been divided
- more things behave correctly (unix style returns and errno)
- prune some functions from source as it's not used
- use meaningful names when using pointers and structs
- remove shared pointer between plugins and prince handler

edoput committed:
add cmake build definition, remove makefile

edoput committed:
create include directories for headers

edoput committed:
add install target to cmake

edoput committed:
switch to custom types for topology, node and prince_handler

edoput committed:
use a common free_<type> approach in all our data structures

edoput committed:
move topology struct and functions to topology module

edoput committed:
format nicely parser.c

edoput committed:
format nicely control flow, add nicer errors and log messages

edoput committed:
refactor prince_handler constructor

edoput committed:
print errors for new socker

edoput committed:
format nicely socket, parser, config modules

edoput committed:
fix return code for main, print error when opening shared library

edoput committed:
refactor config module

- use types
- more error messages
- avoid exiting on unknown configuration keys

edoput committed:
define names for invalid structs and use them

edoput committed:
format nicely functions in plugin interface header

edoput committed:
explicitly state that netjson has been chosen as default

edoput committed:
order headers in oonf library

edoput committed:
refactor topology module

- more error checking on malloc
- format nicely

edoput committed:
indent correctly

edoput committed:
rename module parser to topology_parser

edoput committed:
move plugin type declaration to common header

edoput committed:
add missing get_initial_timers implementation to oonf library

edoput committed:
create load_plugin module

edoput committed:
create prince_handler module

- move structs and constructor/destructor to module
- set up cmake for new module

edoput committed:
create config structs for graph and proto, make new modules

this is very tricky as it requires freeing two more pointers in
prince_handler_t and adding functions to create and free this new
structs

edoput committed:
add graph_config and proto_config when creating a prince_handler

edoput committed:
early return when supplied conf poiter is invalid

edoput committed:
implement loading of plugin functions

edoput committed:
check malloc in new plugin implementation for olsr

edoput committed:
do sanity check in prince main function

edoput committed:
fix get_topology to behave as unix suppose

edoput committed:
free temporary prince_handler in config modules

edoput committed:
fix graph_parser header, format nicely source

edoput committed:
more checks in prince main on bad pointers

edoput committed:
follow unix conventions

- return 0 on success
- be portable

edoput committed:
refactor for error checking

- check for mallocs
- check for return values

edoput committed:
parse netjson and add error messages for unknown keys

edoput committed:
update error check in prince

edoput committed:
move test directory outside of prince

edoput committed:
more error checking in olsr and socket

edoput committed:
remove graph_parser reference

as there is a custom parser for topology data
we don't need to supply it to a plugin

edoput committed:
fix indentation issue

edoput committed:
remove http_receive function as it's not used

edoput committed:
fix new_plugin_p signature

edoput committed:
update readme

edoput committed:
move creation of graph_parser in prince module

this will ensure that the graph parser is always created and freed
from prince constructor and destructor

edoput committed:
add malloc check to brandes module

edoput committed:
move plugin creationg to prince_handler

edoput committed:
remove duplicate file write

gabri94 commented on README.md:
```diff
@@ -1,36 +1,52 @@
 
-Compile and install:
-`make && make install`
+## Unsing Prince with OONF
```
typo

gabri94 commented on prince/conf/conf_c.json:
```diff
@@ -3,7 +3,8 @@
 		"port": 8081 ,
-		"refresh": 1
+		"refresh": 1,
+               "timer_port": 1
```
timer port should be 2009 for oonf

edoput committed:
update prince_handler id from procedure

edoput committed:
implement valid_topo

edoput committed:
update error message for jsoninfo parser

edoput committed:
update find_node to return 0 on success

edoput committed:
add exponential backoff on socket connecting

edoput committed:
handle errors when connecting

this will provide a way to filter on the errors types when
trying to reconnect

remove loading prototype of configurations

edoput committed:
make get_initial_timer behave

edoput committed:
set default timers when none are recovered

edoput committed:
set protocol as olsrv1 in jsoninfo parser

edoput committed:
remove duplicate target

edoput committed:
remove duplicate declaration in brandes
